src/balancerd/src/lib.rs

```diff
@@ -649,13 +649,33 @@ impl PgwireBalancer {
         envd_addr: SocketAddr,
         password: Option<String>,
         params: BTreeMap<String, String>,
-    ) -> Result<TcpStream, anyhow::Error>
+    ) -> Result<Conn<TcpStream>, anyhow::Error>
     where
         A: AsyncRead + AsyncWrite + AsyncReady + Send + Sync + Unpin,
     {
-        let mut mz_stream = TcpStream::connect(envd_addr).await?;
+        let mut tcp_stream = TcpStream::connect(envd_addr).await?;
         let mut buf = BytesMut::new();
 
+        FrontendStartupMessage::SslRequest.encode(&mut buf)?;
+        tcp_stream.write_all(&buf).await?;
+        let mut maybe_ssl_request_response = [0u8; 1];
+        let nread =
+            netio::read_exact_or_eof(&mut tcp_stream, &mut maybe_ssl_request_response).await?;
+        let mut mz_stream = if nread == 1 && maybe_ssl_request_response == [ACCEPT_SSL_ENCRYPTION] {
+            // do a TLS handshake
+            let mut builder =
+                SslConnector::builder(SslMethod::tls()).expect("Error creating builder.");
+            // environmentd doesn't yet have a cert we trust, so for now disable verification.
+            builder.set_verify(SslVerifyMode::NONE);
+            let ssl = builder
+                .build()
+                .configure()?
+                .into_ssl(&envd_addr.to_string())?;
```

Comment:
because you're not using openssl to establish the connection, you have to explicitly call `ssl.set_connect_state()` here to tell it that this should do the ssl handshake as a tls client

Reply:
(see https://docs.openssl.org/master/man3/SSL_set_connect_state/)